File: backend/src/main/java/cn/edu/nju/playground/model/po/Activity.java
package cn.edu.nju.playground.model.po;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import cn.edu.nju.playground.enums.SportsType;
import cn.edu.nju.playground.enums.ActivityStatus;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Data
@EqualsAndHashCode(callSuper = false)
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
@Table(name = "activities", indexes = {
        @Index(name = "idx_activity_type_time", columnList = "type, startTime"),
        @Index(name = "idx_activity_location", columnList = "province, city, district"),
        @Index(name = "idx_activity_status", columnList = "status"),
        @Index(name = "idx_activity_creator", columnList = "creatorId")
})
public class Activity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @JoinColumn(name = "creatorId", nullable = false)
    @ManyToOne(fetch = FetchType.LAZY)
    private User creator;

    @Column(nullable = false)
    private String contactInfo;

    @Column(nullable = false, length = 200)
    private String title;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private SportsType type;

    @Lob
    @Column(nullable = false)
    private String description;

    @ElementCollection
    @Builder.Default
    private List<String> images = List.of();

    @Column(nullable = false)
    private LocalDateTime startTime;

    @Column(nullable = false)
    private LocalDateTime endTime;

    @Column(nullable = false, length = 50)
    private String province;

    @Column(nullable = false, length = 50)
    private String city;

    @Column(nullable = false, length = 50)
    private String district;

    @Column(nullable = false, length = 500)
    private String venue;

    @Column(nullable = false)
    private Integer minParticipants;

    @Column(nullable = false)
    private Integer maxParticipants;

    @Column(nullable = false)
    @Builder.Default
    private Integer currentParticipants = 0;

    @Column(nullable = false, precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal fee = BigDecimal.ZERO;

    @Column(nullable = false)
    private LocalDateTime registrationDeadline;

    @Lob
    private String requirements;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private ActivityStatus status = ActivityStatus.RECRUITING;

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(nullable = false)
    private LocalDateTime updatedAt;
}

File: frontend/src/types/activity.ts
import type { UserPublicInfoResponse, SportsType } from './user'

export type ActivityStatus =
  | 'RECRUITING'
  | 'FULL'
  | 'REGISTRATION_CLOSED'
  | 'IN_PROGRESS'
  | 'COMPLETED'
  | 'CANCELLED'

export type ParticipationType =
  | 'CREATED'
  | 'JOINED'
  | 'APPLIED'
  | 'CANCELLED'

export interface ActivityBriefResponse {
  id: number
  title: string
  type: SportsType
  startTime: string
  endTime: string
  province: string
  city: string
  district: string
  venue: string
  currentParticipants: number
  maxParticipants: number
  fee: number
  status: ActivityStatus
  registrationDeadline: string
  creator: UserPublicInfoResponse
  isRegistered: boolean
  image: string
  createdAt: string
}

export interface ActivityDetailResponse {
  id: number
  title: string
  type: SportsType
  description: string
  images: string[]
  startTime: string
  endTime: string
  province: string
  city: string
  district: string
  venue: string
  minParticipants: number
  maxParticipants: number
  currentParticipants: number
  fee: number
  registrationDeadline: string
  requirements: string
  status: ActivityStatus
  creator: UserPublicInfoResponse
  contactInfo: string
  participants: UserPublicInfoResponse[]
  isRegistered: boolean
  canRegister: boolean
  canUpdate: boolean
  canCancel: boolean
  createdAt: string
  updatedAt: string
}

export interface ActivitySearchParams {
  keyword?: string
  type?: SportsType
  startTime?: string
  endTime?: string
  minFee?: string
  maxFee?: string
  province?: string
  city?: string
  district?: string
  status?: ActivityStatus
  sortBy?: string
  sortDir?: string
  page: number
  size: number
}
